contracts/non-fungible-tokens/andromeda-marketplace/src/contract.rs

```diff
@@ -0,0 +1,946 @@
+        let mut deps = mock_dependencies_custom(&[]);
+        let env = mock_env();
+        let info = mock_info("owner", &[]);
+        let msg = InstantiateMsg { modules: None };
+        let _res = instantiate(deps.as_mut(), env, info, msg).unwrap();
+
+        let hook_msg = Cw721HookMsg::StartSale {
+            coin_denom: "uusd".to_string(),
+            whitelist: None,
+            price: Uint128::zero(),
+        };
+        let msg = ExecuteMsg::ReceiveNft(Cw721ReceiveMsg {
+            sender: MOCK_TOKEN_OWNER.to_owned(),
+            token_id: MOCK_UNCLAIMED_TOKEN.to_owned(),
+            msg: encode_binary(&hook_msg).unwrap(),
+        });
+        let env = mock_env();
+
+        let info = mock_info(MOCK_TOKEN_ADDR, &[]);
+        let err = execute(deps.as_mut(), env, info, msg).unwrap_err();
+        assert_eq!(err, ContractError::InvalidZeroAmount {})
+    }
+
+    #[test]
+    fn execute_buy_with_tax_and_royalty_insufficient_funds() {
+        let mut deps = mock_dependencies_custom(&[]);
+        let env = mock_env();
+        let info = mock_info("owner", &[]);
+        let modules = vec![Module {
+            module_type: RATES.to_owned(),
+            address: AndrAddress {
+                identifier: MOCK_RATES_CONTRACT.to_owned(),
+            },
+            is_mutable: false,
+        }];
+        let msg = InstantiateMsg {
+            modules: Some(modules),
+        };
+        let _res = instantiate(deps.as_mut(), env, info, msg).unwrap();
+
+        start_sale(deps.as_mut(), None);
+        assert_sale_created(deps.as_ref(), None);
+
+        let msg = ExecuteMsg::Buy {
+            token_id: MOCK_UNCLAIMED_TOKEN.to_owned(),
+            token_address: MOCK_TOKEN_ADDR.to_string(),
+        };
+
+        let info = mock_info("someone", &coins(100, "uusd".to_string()));
+        let err = execute(deps.as_mut(), mock_env(), info, msg).unwrap_err();
+        assert_eq!(err, ContractError::InsufficientFunds {})
+    }
+
+    #[test]
+    fn execute_buy_with_tax_and_royalty_works() {
+        let mut deps = mock_dependencies_custom(&[]);
+        let env = mock_env();
+        let info = mock_info("owner", &[]);
+        let modules = vec![Module {
+            module_type: RATES.to_owned(),
+            address: AndrAddress {
+                identifier: MOCK_RATES_CONTRACT.to_owned(),
+            },
+            is_mutable: false,
+        }];
+        let msg = InstantiateMsg {
+            modules: Some(modules),
+        };
+        let _res = instantiate(deps.as_mut(), env, info, msg).unwrap();
+
+        start_sale(deps.as_mut(), None);
+        assert_sale_created(deps.as_ref(), None);
+
+        let msg = ExecuteMsg::Buy {
+            token_id: MOCK_UNCLAIMED_TOKEN.to_owned(),
+            token_address: MOCK_TOKEN_ADDR.to_string(),
+        };
+
+        let info = mock_info("someone", &coins(150, "uusd".to_string()));
+        let _res = execute(deps.as_mut(), mock_env(), info, msg).unwrap();
```

Comment:
I'd verify that the response contains the expected `BankMsg::Send`s to ensure that taxes and royalties are calculated as we expect